Add filter option for observation data

colin-harkins opened the pull request:
Pull request to address issue #146.

This adds a yaml option to filter data more generally. The yaml option lives under the data_proc option in the plotting, statistics and obs sections of the yaml. 

An example of the formatting of this option is: `filter_dict: {'state_name':{'value':['CA','NY'],'oper':'isin'},'WS':{'value':1,'oper':'<'}}`
The above would filter data (airnow here) such that only observations in California or New york where windspeed is < 1 m/s remain.

When `filter_dict` is specified under `data_proc` in a plot group or in the statistics section, the filtering is performed right after the domain filtering. This is done with the pandas query function. 

To filter observations prior to pairing, a `data_proc` section with `filter_dict` specified can be added under the observation type. This will filter the observation data right after the mask and scale operations when reading in observations. This is done with the xarray Dataset.where function, because the observations are stored as an xarray Dataset and I could not figure out how to reproduce the same functionality with the xarray Dataset.query function.

zmoon reviewed (commented):
This is cool, but it seems like it would be simpler and more flexible to support passing in full query strings.

For example,
```
filter_dict: {'state_name':{'value':['CA','NY'],'oper':'isin'},'WS':{'value':1,'oper':'<'}}
```
could become
```yaml
filter: state_name in ['CA', 'NY'] and WS < 1
```
and this string passed to `.query()`.

colin-harkins commented:
I've updated this so that you can filter the data using a string as @zmoon suggested. This only applies to filtering in the plotting and statistics routines. 

The yaml option for this is `filter_string`. I've set it up such that if `filter_dict` is specified in the same `data_proc` section, `filter_string` will be ignored. Does that make most sense or would it be better to let the user specify both?

zmoon commented:
I think it is fine to expect one or the other, not both. And maybe raise error or warning if both are set.

colin-harkins commented:
I modified the behavior so that it just raises an error when both are set. I also changed the docs accordingly. I feel like that's probably safer than raising a warning that the user could potentially miss.

rschwant commented on docs/appendix/yaml.rst:
```diff
@@ -285,7 +300,25 @@ observation label is first and the model label is second
+     to the plotting routine. The result is the same as using filter_dict.
+     This uses the pandas query method on the paired dataset.
+     This cannot be specified if 'filter_dict' is specified.
+     Example: state_name in ['CO'] and WS < 1
```
Can we add something like the following?: "This option is only available for for surface and aircraft observations. For satellite observations, use the filter_dict option instead."

rschwant commented on docs/appendix/yaml.rst:
```diff
@@ -360,7 +393,25 @@ observation label is first and the model label is second
+     to the statistics routine. The result is the same as using filter_dict.
+     This uses the pandas query method on the paired dataset.
+     This cannot be specified if 'filter_dict' is specified.
+     Example: state_name in ['CO'] and WS < 1
```
Can we add something like the following?: "This option is only available for for surface and aircraft observations. For satellite observations, use the filter_dict option instead."

rschwant reviewed (commented):
Colin, this looks great! I tested it and it works well. I'm glad we added the filter_string option because you can do more complex filters that way like I wanted to test restricting WS < 1 m/s or WS is a NaN and with the filter_string option you can do this. I just suggested a couple small documentation updates and if you agree, I'll approve. Thanks!

colin-harkins commented:
@rschwant I've updated this with all of the suggestions you made.

rschwant reviewed (approved):
Great thanks, I approve, go ahead and merge it in.
